### PR TITLE
bpo-41520: codeop no longer ignores SyntaxWarning

### PR DESCRIPTION
**Pull request by vstinner**
https://bugs.python.org/issue41520

### CODE REVIEW

**Comment by vstinner** [comment]
cc @csabella @terryjreedy @Carreau @1st1

File: Lib/codeop.py [hunk 1]
```diff
@@ -84,9 +84,11 @@ def _maybe_compile(compiler, source, filename, symbol):
     except SyntaxError:
         pass
 
-    # Suppress warnings after the first compile to avoid duplication.
+    # Catch syntax warnings after the first compile
+    # to emit SyntaxWarning at most once.
```
Comment (vstinner) [inline Lib/codeop.py:88]: Hum, english is not my first language. I would like to say "the warning must be emitted 0 or 1 time, but not more". Is it ok to say "at most once"?

Comment (csabella) [reply Lib/codeop.py:88]: I think "at most once" would be correct since "at least once" would mean one or more time.

**Review by csabella — APPROVED** [review]
LGTM.  @vstinner, thank you for fixing this.  🙇‍♀️

**Comment by csabella** [comment]
@Carreau, thank you for catching this.

**Comment by vstinner** [comment]
Thanks for the review @csabella!

File: Lib/codeop.py [hunk 2]
```diff
@@ -84,9 +84,11 @@ def _maybe_compile(compiler, source, filename, symbol):
...
     with warnings.catch_warnings():
-        warnings.simplefilter("ignore")
+        warnings.simplefilter("error", SyntaxWarning)
```
Comment (terryjreedy) [inline Lib/codeop.py:90]: Ignoring *all* warnings was intentional.  Compile also issues DeprecationWarnings.  Narrowing the filter to SyntaxWarning only reintroduces the error of DeprecationWarning being issued thrice, which was fixed by the previous code.  I will write more on the issue.

Followup PR fixes regression by removing SyntaxWarning.

File: Lib/test/test_codeop.py [hunk 3]
```diff
@@ -310,5 +311,11 @@ def test_warning(self):
             compile_command("0 is 0")
             self.assertEqual(len(w.warnings), 1)
 
+        # bpo-41520: check SyntaxWarning treated as an SyntaxError
+        with self.assertRaises(SyntaxError):
+            warnings.simplefilter('error', SyntaxWarning)
+            compile_command('1 is 1\n', symbol='exec')
```
Comment (terryjreedy) [inline Lib/test/test_codeop.py:317]: This source should not have '\n' as single lines are passed without it.  That is why to recompile with \n and \n\n added.  Fixed in followup PR.

Comment (Carreau) [reply Lib/test/test_codeop.py:317]: Even w/o the `\n` 3.8.5 does not raise the SyntaxError.

```
$ cat foo.py
import warnings
from codeop import compile_command

warnings.simplefilter('error', SyntaxWarning)
res = compile_command('1 is 1', symbol='exec')
print('Res', res)

$ python foo.py
Res None
```

File: Lib/codeop.py (hunk 2 above)
Comment (Carreau) [reply Lib/codeop.py:90]: Then maybe the catch warnings need to be put in there 3 times, to reset the warning filter on each subsequent compilations ?

**Comment by terryjreedy** [comment]
The title is misleading because before this patch, codeop was not ignoring the first SyntaxWarning.  The test we added in bpo-40807 ensured that.  The issue, rather, is that _maybe_compile requires 3 SyntaxErrors, not just 1, to raise a SyntaxError.  As I note above, the fix to a regression introduces another regression.  Fixed in #21848.

File: Lib/test/test_codeop.py (hunk 3 above)
Comment (terryjreedy) [reply Lib/test/test_codeop.py:317]: I expect that, but aside from your issue, the test if better if it calls compile_command the way it is intended to be used and is used by code.InteractiveInterpreter.

**Pull request merged**